DDG4/edm4hep/EDM4hepFileReader.cpp

```diff
@@ -0,0 +1,332 @@
+        } catch(std::exception& ) {}
+        // Create input event parameters context
+        try {
+          Geant4Context* ctx = context();
+          EventParameters *parameters = new EventParameters();
+          parameters->setRunNumber(runNumber);
+          parameters->setEventNumber(eventNumber);
+          parameters->ingestParameters(frame.getParameters());
+          ctx->event().addExtension<EventParameters>(parameters);
+        } catch(std::exception &) {}
+      } else {
+        printout(WARNING,"EDM4hepFileReader","No EventHeader collection found or too many event headers!");
+        try {
+          Geant4Context* ctx = context();
+          EventParameters *parameters = new EventParameters();
+          parameters->setRunNumber(0);
+          parameters->setEventNumber(event_number);
+          parameters->ingestParameters(frame.getParameters());
+          ctx->event().addExtension<EventParameters>(parameters);
+        } catch(std::exception &) {}
+      }
+      printout(INFO,"EDM4hepFileReader","read collection %s from event %d in run %d ",
+               m_collectionName.c_str(), eventNumber, runNumber);
+
+    } else {
+      return EVENT_READER_EOF;
+    }
+
+    printout(INFO,"EDM4hepFileReader", "We read the particle collection");
+    int NHEP = primaries.size();
+    // check if there is at least one particle
+    if ( NHEP == 0 ) {
+      printout(WARNING,"EDM4hepFileReader", "We have no particles");
+      return EVENT_READER_NO_PRIMARIES;
+    }
+
+    MCPARTICLE_MAP mcparts;
+    std::vector<int>  mcpcoll;
+    mcpcoll.resize(NHEP);
+    for(int i=0; i<NHEP; ++i ) {
+      edm4hep::MCParticle p = primaries.at(i);
+      mcparts[p.getObjectID().index] = i;
+      mcpcoll[i] = p.getObjectID().index;
+    }
+
+    // build collection of MCParticles
+    for(int i=0; i<NHEP; ++i )   {
+      const auto& mcp = primaries.at(i);
+      Geant4ParticleHandle p(new Particle(i));
+      const auto mom   = mcp.getMomentum();
+      const auto vsx   = mcp.getVertex();
+      const auto vex   = mcp.getEndpoint();
+      const auto spin  = mcp.getSpin();
+      const int  pdg   = mcp.getPDG();
+      p->pdgID        = pdg;
+      p->charge       = int(mcp.getCharge()*3.0);
+      p->psx          = mom[0]*CLHEP::GeV;
+      p->psy          = mom[1]*CLHEP::GeV;
+      p->psz          = mom[2]*CLHEP::GeV;
+      p->time         = mcp.getTime()*CLHEP::ns;
+      p->properTime   = mcp.getTime()*CLHEP::ns;
+      p->vsx          = vsx[0]*CLHEP::mm;
+      p->vsy          = vsx[1]*CLHEP::mm;
+      p->vsz          = vsx[2]*CLHEP::mm;
+      p->vex          = vex[0]*CLHEP::mm;
+      p->vey          = vex[1]*CLHEP::mm;
+      p->vez          = vex[2]*CLHEP::mm;
+      p->process      = 0;
+      p->spin[0]      = spin[0];
+      p->spin[1]      = spin[1];
+      p->spin[2]      = spin[2];
+      p->colorFlow[0] = 0;
+      p->colorFlow[1] = 0;
```

Comment:
We need to drop the `colorFlow` here as https://github.com/key4hep/EDM4hep/pull/389 has been merged. Additionally `spin` will be changed as well, but we can get to that once I have had time to do that.

Reply:
Right, too many things with the same members flying around here. Sorry got confused.